api/src/main/java/org/apache/iceberg/expressions/ExpressionUtil.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.expressions;
+
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import org.apache.iceberg.transforms.Transform;
+import org.apache.iceberg.transforms.Transforms;
+import org.apache.iceberg.types.Types;
+
+/**
+ * Expression utility methods.
+ */
+public class ExpressionUtil {
+  private static final Transform<CharSequence, Integer> HASH_FUNC = Transforms
+      .bucket(Types.StringType.get(), Integer.MAX_VALUE);
+  private static final Pattern DATE = Pattern.compile("\\d\\d\\d\\d-\\d\\d-\\d\\d");
+  private static final Pattern TIME = Pattern.compile(
+      "\\d\\d:\\d\\d(:\\d\\d(.\\d{1,6})?)?");
+  private static final Pattern TIMESTAMP = Pattern.compile(
+      "\\d\\d\\d\\d-\\d\\d-\\d\\dT\\d\\d:\\d\\d(:\\d\\d(.\\d{1,6})?)?([-+]\\d\\d:\\d\\d)?");
+
+  private ExpressionUtil() {
+  }
+
+  /**
+   * Produces an unbound {@link Expression} with the same structure, but with data values replaced by descriptions.
+   * <p>
+   * Numbers are replaced with magnitude and type, string-like values are replaced by hashes, and date/time values are
+   * replaced by the type.
+   *
+   * @param expr an Expression to sanitize
+   * @return a sanitized Expression
+   */
+  public static Expression sanitize(Expression expr) {
+    return ExpressionVisitors.visit(expr, ExpressionSanitizer.INSTANCE);
+  }
+
+  /**
+   * Produces a sanitized expression string with the same structure, but with data values replaced by descriptions.
+   * <p>
+   * Numbers are replaced with magnitude and type, string-like values are replaced by hashes, and date/time values are
+   * replaced by the type.
+   *
+   * @param expr an Expression to sanitize
+   * @return a sanitized expression string
+   */
+  public static String toSanitizedString(Expression expr) {
```

Comment:
As the specific needs for sanitization are often very business specific, would it make sense to leave this somehow as something that end users could override somewhat easily?

Possibly the `santitize` method itself could be abstract with this as the default implementation.

Reply:
We can improve this as needed, but I don't want to guess ahead of time what makes sense. People can always extend this later.